File: src/unity/Runtime/Ads/Internal/GuardedInterstitialAd.cs
using System.Threading.Tasks;

namespace EE.Internal {
    internal class GuardedInterstitialAd
        : ObserverManager<IInterstitialAdObserver>, IInterstitialAd {
        private readonly IInterstitialAd _ad;
        private bool _loading;
        private bool _displaying;
        private readonly ObserverHandle _handle;
        private readonly Retrier _retrier;

        public GuardedInterstitialAd(IInterstitialAd ad) {
            _ad = ad;
            _handle = new ObserverHandle();
            _handle.Bind(ad).AddObserver(new IInterstitialAdObserver {
                OnLoaded = () => DispatchEvent(observer => observer.OnLoaded?.Invoke()),
                OnClicked = () => DispatchEvent(observer => observer.OnClicked?.Invoke())
            });
            _retrier = new Retrier(1, 2, 64);
        }

        public void Destroy() {
            _ad.Destroy();
            _handle.Clear();
        }

        public bool IsLoaded { get; private set; }

        public async Task<bool> Load() {
            if (IsLoaded) {
                return true;
            }
            if (_displaying) {
                return false;
            }
            if (_loading) {
                return false;
            }
            _loading = true;
            IsLoaded = await _ad.Load();
            if (IsLoaded) {
                _retrier.Stop();
            } else {
                Utils.NoAwait(async () => {
                    await _retrier.Process(async () => { //
                        return IsLoaded = await _ad.Load();
                    });
                });
            }
            _loading = false;
            return IsLoaded;
        }

        public async Task<bool> Show() {
            if (!IsLoaded) {
                return false;
            }
            if (_loading) {
                return false;
            }
            if (_displaying) {
                return false;
            }
            _displaying = true;
            IsLoaded = false;
            var result = await _ad.Show();
            _displaying = false;
            return result;
        }
    }
}

File: src/cpp/ee/ads/internal/GuardedInterstitialAd.cpp
#include "ee/ads/internal/GuardedInterstitialAd.hpp"

#include <ee/core/NoAwait.hpp>
#include <ee/core/ObserverHandle.hpp>
#include <ee/core/Task.hpp>
#include <ee/core/Utils.hpp>

#include "ee/ads/internal/Retrier.hpp"

namespace ee {
namespace ads {
using Self = GuardedInterstitialAd;

Self::GuardedInterstitialAd(const std::shared_ptr<IInterstitialAd>& ad)
    : ad_(ad) {
    loading_ = false;
    loaded_ = false;
    displaying_ = false;

    handle_ = std::make_unique<ObserverHandle>();
    handle_->bind(*ad_).addObserver({
        .onLoaded =
            [this] {
                // Propagation.
                dispatchEvent([](auto&& observer) {
                    if (observer.onLoaded) {
                        observer.onLoaded();
                    }
                });
            },
        .onClicked =
            [this] {
                // Propagation.
                dispatchEvent([](auto&& observer) {
                    if (observer.onClicked) {
                        observer.onClicked();
                    }
                });
            },
    });

    retrier_ = std::make_unique<Retrier>(1, 2, 64);
}

Self::~GuardedInterstitialAd() = default;

void Self::destroy() {
    ad_->destroy();
    handle_->clear();
    retrier_->stop();
}

bool Self::isLoaded() const {
    return loaded_;
}

Task<bool> Self::load() {
    if (loaded_) {
        co_return true;
    }
    if (displaying_) {
        co_return false;
    }
    if (loading_) {
        co_return false;
    }
    loading_ = true;
    loaded_ = co_await ad_->load();
    if (loaded_) {
        retrier_->stop();
    } else {
        noAwait([this]() -> Task<> {
            co_await retrier_->process([this]() -> Task<bool> {
                co_return loaded_ = co_await ad_->load();
            });
        });
    }
    loading_ = false;
    co_return loaded_;
}

Task<bool> Self::show() {
    if (not loaded_) {
        co_return false;
    }
    if (loading_) {
        co_return false;
    }
    if (displaying_) {
        co_return false;
    }
    displaying_ = true;
    loaded_ = false;
    auto result = co_await ad_->show();
    displaying_ = false;
    co_return result;
}
} // namespace ads
} // namespace ee
